### PR TITLE
add test that check manual bootstrap token

### PR DESCRIPTION
**Pull request by mgold1234**
related to those issues -#201 , #190

### CODE REVIEW

**Comment by qiujian16** [comment]
@mgold1234 I think you can discuss with @ycyaoxdu on how it can be tested. @ycyaoxdu was mainly working on the e2e tests.

File: test/e2e/clusteradm/joinhubscenario_skip_approve_test.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+						"init/bootstrap_sa_cluster_role_binding_manual_token.yaml",
+					)
+					_, err = applier.ApplyDirectly(reader, nil, false, "", file1...)
+					gomega.Expect(err).To(gomega.BeNil())
+					sa, err:= kubeClient.CoreV1().
+						ServiceAccounts(config.OpenClusterManagementNamespace).
+						Get(context.TODO(), "sa-manual-token", metav1.GetOptions{})
+					fmt.Println("ssssserviceacoountssssssssecret", sa)
+					if err!=nil{
+						return
```
Comment (itdove) [inline test/e2e/clusteradm/joinhubscenario_skip_approve_test.go:61]: Please use here something like "gomega.Expect(err).To(gomega.BeNil())"

File: test/e2e/clusteradm/joinhubscenario_skip_approve_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+					applier := applierBuilder.WithClient(kubeClient, apiExtensionsClient, dynamicClient).Build()
+					_, err = applier.ApplyDirectly(reader, nil, false, "", files...)
+					gomega.Expect(err).To(gomega.BeNil())
+					time.Sleep(1000000)
+					file1 := []string{}
+					file1 = append(file1,
+					"init/bootstrap_cluster_role.yaml",
+						"init/bootstrap_sa_cluster_role_binding_manual_token.yaml",
+					)
+					_, err = applier.ApplyDirectly(reader, nil, false, "", file1...)
```
Comment (itdove) [inline test/e2e/clusteradm/joinhubscenario_skip_approve_test.go:54]: i never tried with nil, can you try with values := make(map[string]interface{})

File: test/e2e/clusteradm/joinhubscenario_skip_approve_test.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+					file1 = append(file1,
+					"init/bootstrap_cluster_role.yaml",
+						"init/bootstrap_sa_cluster_role_binding_manual_token.yaml",
+					)
+					_, err = applier.ApplyDirectly(reader, make(map[string]interface{}), false, "", file1...)
+					gomega.Expect(err).To(gomega.BeNil())
+					sa, err:= kubeClient.CoreV1().
+						ServiceAccounts(config.OpenClusterManagementNamespace).
+						Get(context.TODO(), "sa-manual-token", metav1.GetOptions{})
+					fmt.Println("ssssserviceacoountssssssssecret", sa)
```
Comment (ycyaoxdu) [inline test/e2e/clusteradm/joinhubscenario_skip_approve_test.go:59]: do we need this Println here?

File: test/e2e/clusteradm/joinhubscenario_skip_approve_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+					files = append(files,
+						"init/bootstrap_sa_manual_token.yaml",
+					)
+
+					reader := scenario.GetScenarioResourcesReader()
+					applierBuilder := &apply.ApplierBuilder{}
+					applier := applierBuilder.WithClient(kubeClient, apiExtensionsClient, dynamicClient).Build()
+					_, err = applier.ApplyDirectly(reader, make(map[string]interface{}), false, "", files...)
+					gomega.Expect(err).To(gomega.BeNil())
+					time.Sleep(1000000)
```
Comment (ycyaoxdu) [inline test/e2e/clusteradm/joinhubscenario_skip_approve_test.go:48]: can sleep time be shorter?

File: test/e2e/util/util.go [hunk 5]
```diff
@@ -19,6 +21,34 @@ func PrepareE2eEnvironment() *TestE2eConfig {
...
+	if v := os.Getenv("MANAGED_CLUSTER2_NAME"); v == "" {
+		os.Setenv("MANAGED_CLUSTER2_NAME", projectName+"-e2e-test-c2")
+	}
+	if v := os.Getenv("MANAGED_CLUSTER2_CTX"); v == "" {
+		os.Setenv("MANAGED_CLUSTER2_CTX", "kind-"+os.Getenv("MANAGED_CLUSTER2_NAME"))
+	}
+
+	home, err := os.UserHomeDir()
+	Expect(err).To(BeNil())
+	os.Setenv("KUBECONFIG", filepath.Join(home,".kube","config"))
```
Comment (ycyaoxdu) [inline test/e2e/util/util.go:49]: already set env in makefile https://github.com/open-cluster-management-io/clusteradm/blob/main/test/e2e/e2e-test.mk#L2-L10

File: test/e2e/clusteradm/joinhubscenario_skip_approve_test.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+				}
+				if foundSecret == nil {
+					return fmt.Errorf("Secret with prefix %s not found, trying again", prefix)
+				}
+				return nil
+			}, 30, 1).Should(gomega.BeNil())
+
+			gomega.Expect(sa.Name).To(gomega.Equal("sa-manual-token"))
+			fmt.Println("service account", sa)
+			fmt.Println("secret", foundSecret)
```
Comment (ycyaoxdu) [inline test/e2e/clusteradm/joinhubscenario_skip_approve_test.go:84]: remove print

File: test/e2e/clusteradm/joinhubscenario_skip_approve_test.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+			token := string(foundSecret.Data["token"])
+
+			fmt.Println("tokeeeeeeeeeeeeeeeeen", token)
```
Comment (ycyaoxdu) [inline test/e2e/clusteradm/joinhubscenario_skip_approve_test.go:87]: remove print

**Comment by itdove** [comment]
@ycyaoxdu As for the time being with the makefile we can not run on debug mode and so very difficult to debug.
Also we added a focus, so the developer has a chance to run only 1 Description rather than the whole lot and have wait a lot of time to see the results.

File: test/e2e/clusteradm/e2e_suite_test.go [hunk 8]
```diff
@@ -11,14 +22,52 @@ import (
...
+	dynamicClient, err = dynamic.NewForConfig(hubConfig)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	clusterClient, err = clusterv1client.NewForConfig(hubConfig)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+	clusterv1.AddToScheme(scheme.Scheme)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+	restConfig = hubConfig
+})
```
Comment (ycyaoxdu) [inline test/e2e/clusteradm/e2e_suite_test.go:73]: eof

File: test/e2e/clusteradm/joinhubscenario_skip_approve_test.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+			err = e2e.Clusteradm().Accept(
+				"--clusters", e2e.Cluster().ManagedCluster1().Name(),
+				"--wait", "30",
+				"--context", e2e.Cluster().Hub().Context(),
+				"--skip-approve-check",
+			)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred(), "clusteradm accept error")
+		})
+	})
+})
```
Comment (ycyaoxdu) [inline test/e2e/clusteradm/joinhubscenario_skip_approve_test.go:102]: eof

**Review by ycyaoxdu — COMMENTED** [review]
/lgtm

**Pull request merged**